### PR TITLE
feat: force packing if threshold is not reached after a configurable duration

### LINKED ISSUE #169: Last pack job should get prepared after a certain duration
In the current implementation, only when the size of a pack job is more than a threshold (default 16GB), it will be prepared by the dataset worker.

This means if the pack job never reaches 16GB, it will never get prepared, which can happen if the client has less than 16GB data or their last chunk of data is less than 16GB and they are no longer pushing new blobs.

The fix is to pack remaining files after a certain duration (24h, hardcoded or configurable)

### PR DESCRIPTION
**Pull request by elijaharita**
the default duration is 24h.

this PR doesn't introduce a test (yet?). i've only verified its function by looking at logs so far.

i'm thinking about how this could be tested. maybe could add a routine to the beginning of the integration test file that ensures no pack jobs have run yet, then post a small amount of data below the threshold, wait a bit longer than the configured timeout (set to 5 seconds for the test, or something like that), and then check that a pack job has run.

thoughts? i haven't looked into whether it's possible / how to inspect pack jobs yet

Fixes #169

### CODE REVIEW

**Review by willscott — COMMENTED** [review]
* the most common pattern for testing is that you take i a 'clock' object, and for the tests have a mock implementation that can jump ahead to later times as part of the test.
* rather than packing as a timeout, we have considered, and it may be worthwhile to have an exposed 'flush' endpoint to allow a client to force packing to occur and indicate they're done with a dataset and would like to immediately force all unsaved state out to durable storage - e.g. before shutting down the docker compose instance.

File: integration/singularity/store.go [hunk 1]
```diff
@@ -290,31 +293,51 @@ func (s *SingularityStore) runPreparationJobs() {
...
-				if err != nil {
+			if prepareToPackFileRes.Payload > s.packThreshold {
+				if err := s.prepareToPackSource(ctx); err != nil {
 					logger.Errorw("preparing to pack source", "error", err)
 				}
 			}
+
+		// If forced pack message comes through (e.g. from pack threshold max
+		// wait time being exceeded), prepare to pack source immediately
+		case <-s.forcePack.C:
```
Comment (xinaxu) [inline integration/singularity/store.go:318]: We need to handle a few more cases here:
1. If new files keep coming and new pack jobs are being marked as ready, we should extend the forcePack timer
2. If new files keep coming but they are so small and won't fill up a packJob for days, should we do forcePack (I'm leaning toward yes)

**Review by xinaxu — CHANGES_REQUESTED** [review]

**Comment by xinaxu** [comment]
> * rather than packing as a timeout, we have considered, and it may be worthwhile to have an exposed 'flush' endpoint to allow a client to force packing to occur and indicate they're done with a dataset and would like to immediately force all unsaved state out to durable storage - e.g. before shutting down the docker compose instance.

I think the "flush" API makes DeStore less simple than it should be, as it exposes some level of internal implmentation to user. They can, alternatively, use internal Singularity API to achieve the same goal so I don't strongly feel we need this "flush" API now.
https://data-programs.gitbook.io/singularity/web-api-reference/job#preparation-id-source-name-finalize

File: integration/singularity/store.go (hunk 1 above)
Comment (elijaharita) [reply integration/singularity/store.go:318]: also, what would be a good default value?

i'd probably rename PACK_THRESHOLD_MAX_WAIT to FORCE_PACK_AFTER

```
FORCE_PACK_AFTER=24h
FORCE_PACK_BELOW_HOURLY_RATE=256MiB
```

?

Comment (xinaxu) [reply integration/singularity/store.go:318]: My opinion is to just implement FORCE_PACK_AFTER if there is no packJob being marked for last X hours. Hourly rate is a good idea but I would defer it until we hear some feedback from Beta users.
Curious about what others think. @masih @willscott

Comment (elijaharita) [reply integration/singularity/store.go:319]: ok, updated with that behavior for now.

**Review by xinaxu — APPROVED** [review]

**Pull request merged**